added possible solution to compiler error message

tudorlesan opened the pull request:
Fixes #220 
Fixes #222
Signed-off-by: Lesan Tudor <tudor-andrei.lesan@hp.com>

tudorlesan commented:
Since #220 and #222 are not more detailed, I think this solves both issues.

meirwah commented on cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/parser/YamlParser.java:
```diff
@@ -39,7 +41,15 @@ public ParsedSlang parse(SlangSource source) {
+            if (e instanceof ScannerException && errorMessage.startsWith("mapping values")){
+                errorMessage += "Probably did not provide (key: value) pair or missing space after colon(:)";
+            }
+            if (e instanceof ConstructorException && errorMessage.startsWith("Cannot create property")){
```
why not "else if"?

tudorlesan replied:
I figured "e" cannot be two error types at once. But I can add "else if" just to be more clear

meirwah commented on cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/parser/YamlParser.java:
```diff
@@ -39,7 +41,15 @@ public ParsedSlang parse(SlangSource source) {
-            throw new RuntimeException("There was a problem parsing the YAML source: " + source.getName() + ".\n" + e.getMessage(), e);
+            String errorMessage = e.getMessage();
+            if (e instanceof ScannerException && errorMessage.startsWith("mapping values")){
+                errorMessage += "Probably did not provide (key: value) pair or missing space after colon(:)";
```
errorMessage +=  ?
I think the idea is that you write the specific error msg ("Probably did not provid..") before the general one no?

tudorlesan replied:
Well I was thinking the actual problem was presented first, then a possible solution could be printed. Like in C# there is the error message: The type or namespace name 'a_namespace' could not be found (are you missing a using directive or an assembly reference?).

oritstone replied:
I'm OK with this order

oritstone commented on cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/parser/YamlParser.java:
```diff
@@ -39,7 +41,15 @@ public ParsedSlang parse(SlangSource source) {
+                errorMessage += "Probably did not provide (key: value) pair or missing space after colon(:)";
+            }
+            else if (e instanceof ConstructorException && errorMessage.startsWith("Cannot create property")){
+                errorMessage += "Are you sure you are not missing a space after colon(:) for one of the imports and that all imports have an alias?";
```
The exception message starts with: *Cannot create property=imports* and I think you should also the imports text when checking, since you only address errors in imports...

tudorlesan replied:
I did not fully understand this. Could you elaborate please?

oritstone replied:
Your error message is specific to imports. If this is the case, you should include the property name (imports) in the check of the exception message ("Cannot create property=imports").
In your tests you tested a more general use case, where the error could come from any property for which a space is missing. So maybe it is better to leave the if statement the same, but make the error message more general. Not only for the imports use case

oritstone commented:
Missing unit tests for these test cases. Add tests to: CompilerErrorsTest

oritstone commented:
This fix actually deals only with errors in the first alias of the imports.
If the problem is with the second alias, such as:
```yaml
imports:
  consul: io.cloudslang.consul
  consul2:aa
``` 
Then your fix does not catch it.

Not sure we have a way to catch it, but please check

tudorlesan commented:
The same exception (ScannerException) is thrown in that case only it contains a different message. Just had to include that message in the condition and it catches it

oritstone commented on cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/parser/YamlParser.java:
```diff
@@ -39,7 +41,21 @@ public ParsedSlang parse(SlangSource source) {
+                    errorMessage += "Probably did not provide (key: value) pair or missing space after colon(:)";
+                }
+            }
+            throw new RuntimeException("There was a problem parsing the YAML source: " + source.getName() + ".\n" + errorMessage, e);
```
@tudorlesan In the message thrown in the runtime exception, the exception error message appears twice - once from the e.getMessage() and once from the errorMessage.
You should only throw the exception message once.

orius123 commented:
@tudorlesan can you please extract constant values and add comments to the code. 
Since there is a lot of searching through strings here it's really hard to follow from someone that is out of context (like me..)

tudorlesan commented:
@orius123 I extracted the constant values and added some comments on the code, tests, and test flows to associate tests, testflows and errors. Basically there are 4 types of errors covered in this PR. The code checks each error message to establish what specific message to add. 3 of them contain the same message being added, while for the other, a parsing is made to be able to print what property is undefined. Each testflow now also has a comment saying what the problem is. Tell me if you still have questions.

orius123 commented:
@tudorlesan it looks great.
just one small OCD comment.. can you please make the tests use the same constants please.. :grin:
